storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,669 @@
+            }
+        }
+    }
+
+    protected Map<Integer, LocalAssignment> getReassignExecutors(Map<Integer, LocalAssignment> assignExecutors, Set<Integer> keepPorts) {
+        Map<Integer, LocalAssignment> reassignExecutors = new HashMap<>();
+        reassignExecutors.putAll(assignExecutors);
+        for (Integer port : keepPorts) {
+            reassignExecutors.remove(port);
+        }
+        return reassignExecutors;
+    }
+
+
+
+    /**
+     * Returns map from worker id to worker heartbeat. if the heartbeat is nil, then the worker is dead
+     * 
+     * @param assignedExecutors
+     * @return
+     * @throws Exception
+     */
+    public Map<String, StateHeartbeat> getLocalWorkerStats(SupervisorData supervisorData, Map<Integer, LocalAssignment> assignedExecutors, int now) throws Exception {
+        Map<String, StateHeartbeat> workerIdHbstate = new HashMap<>();
+        Map conf = supervisorData.getConf();
+        LocalState localState = supervisorData.getLocalState();
+        Map<String, LSWorkerHeartbeat> idToHeartbeat = SupervisorUtils.readWorkerHeartbeats(conf);
+        Map<String, Integer> approvedWorkers = localState.getApprovedWorkers();
+        Set<String> approvedIds = new HashSet<>();
+        if (approvedWorkers != null) {
+            approvedIds.addAll(approvedWorkers.keySet());
+        }
+        for (Map.Entry<String, LSWorkerHeartbeat> entry : idToHeartbeat.entrySet()) {
+            String workerId = entry.getKey();
+            LSWorkerHeartbeat whb = entry.getValue();
+            State state;
+            if (whb == null) {
+                state = State.notStarted;
+            } else if (!approvedIds.contains(workerId) || !matchesAssignment(whb, assignedExecutors)) {
+                state = State.disallowed;
+            } else if (supervisorData.getDeadWorkers().contains(workerId)) {
+                LOG.info("Worker Process {}as died", workerId);
+                state = State.timedOut;
+            } else if (SupervisorUtils.isWorkerHbTimedOut(now, whb, conf)) {
+                state = State.timedOut;
+            } else {
+                state = State.valid;
+            }
+            LOG.debug("Worker:{} state:{} WorkerHeartbeat:{} at supervisor time-secs {}", workerId, state, whb, now);
+            workerIdHbstate.put(workerId, new StateHeartbeat(state, whb));
+        }
+        return workerIdHbstate;
+    }
+
+    protected boolean matchesAssignment(LSWorkerHeartbeat whb, Map<Integer, LocalAssignment> assignedExecutors) {
+        LocalAssignment localAssignment = assignedExecutors.get(whb.get_port());
+        if (localAssignment == null || !localAssignment.get_topology_id().equals(whb.get_topology_id())) {
+            return false;
+        }
+        List<ExecutorInfo> executorInfos = new ArrayList<>();
+        executorInfos.addAll(whb.get_executors());
+        // remove SYSTEM_EXECUTOR_ID
+        executorInfos.remove(new ExecutorInfo(-1, -1));
```

Comment:
Very minor nit. Can we move this ExecutorInfo to Contants as something like
```
public static final SYSTEM_EXECUTOR_INFO = new ExecutorInfo(-1, -1);
```

so when worker.clj goes to java we can drop SYSTEM-EXECUTOR_ID from Constants?